### ISSUE TITLE
ISE: Cannot use getDescription if this dirty resource is not connected to a document

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-eclipse#1139]
we observer following ISE when deleting a project with a dirty open file

```
	java.lang.IllegalStateException: Cannot use getDescription if this dirty resource is not connected to a document
	at org.eclipse.xtext.ui.editor.DocumentBasedDirtyResource.getDescription(DocumentBasedDirtyResource.java:135)
	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport.descriptionsChanged(DirtyStateEditorSupport.java:485)
	at org.eclipse.xtext.resource.impl.AbstractResourceDescriptionChangeEventSource.notifyListeners(AbstractResourceDescriptionChangeEventSource.java:52)
	at org.eclipse.xtext.ui.notification.StateChangeEventBroker.descriptionsChanged(StateChangeEventBroker.java:20)
	at org.eclipse.xtext.resource.impl.AbstractResourceDescriptionChangeEventSource.notifyListeners(AbstractResourceDescriptionChangeEventSource.java:52)
	at org.eclipse.xtext.builder.impl.DirtyStateAwareResourceDescriptions.globalDescriptionsChanged(DirtyStateAwareResourceDescriptions.java:98)
	at org.eclipse.xtext.builder.impl.DirtyStateAwareResourceDescriptions$GlobalStateListener.descriptionsChanged(DirtyStateAwareResourceDescriptions.java:114)
	at org.eclipse.xtext.resource.impl.AbstractResourceDescriptionChangeEventSource.notifyListeners(AbstractResourceDescriptionChangeEventSource.java:52)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.update(AbstractBuilderState.java:159)
	at org.eclipse.xtext.builder.impl.ProjectOpenedOrClosedListener.processClosedProjects(ProjectOpenedOrClosedListener.java:282)
	at org.eclipse.xtext.builder.impl.ProjectOpenedOrClosedListener$RemoveProjectsJob.runInWorkspace(ProjectOpenedOrClosedListener.java:104)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:42)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

in the debugger i can see that

```
JvmTypesAwareDirtyStateEditorSupport(DirtyStateEditorSupport).descriptionsChanged(IResourceDescription$Event) line: 482	
StateChangeEventBroker(AbstractResourceDescriptionChangeEventSource).notifyListeners(IResourceDescription$Event) line: 52	
StateChangeEventBroker.descriptionsChanged(IResourceDescription$Event) line: 20	
DirtyStateAwareResourceDescriptions(AbstractResourceDescriptionChangeEventSource).notifyListeners(IResourceDescription$Event) line: 52	
DirtyStateAwareResourceDescriptions.globalDescriptionsChanged(IResourceDescription$Event) line: 98	
DirtyStateAwareResourceDescriptions$GlobalStateListener.descriptionsChanged(IResourceDescription$Event) line: 114	
ClusteringBuilderState93k(AbstractResourceDescriptionChangeEventSource).notifyListeners(IResourceDescription$Event) line: 52	
ClusteringBuilderState93k(AbstractBuilderState).update(BuildData, IProgressMonitor) line: 159	
ProjectOpenedOrClosedListener.processClosedProjects(IProgressMonitor) line: 282	
ProjectOpenedOrClosedListener$RemoveProjectsJob.runInWorkspace(IProgressMonitor) line: 104	
ProjectOpenedOrClosedListener$RemoveProjectsJob(InternalWorkspaceJob).run(IProgressMonitor) line: 42	

```

and

```PersistentDataAwareDirtyResource(DocumentBasedDirtyResource).disconnect(IXtextDocument) line: 64	
JvmTypesAwareDirtyStateEditorSupport(DirtyStateEditorSupport).removeDirtyStateSupport(DirtyStateEditorSupport$IDirtyStateEditorSupportClient) line: 466	
XXXEditor(XtextEditor).dispose() line: 584	
XXXEditor.dispose() line: 36	
MultiPageEditorPart$4.run() line: 969	
SafeRunner.run(ISafeRunnable) line: 45	
XXXMultipageEditor(MultiPageEditorPart).disposePart(IWorkbenchPart) line: 965	
XXXlMultipageEditor(MultiPageEditorPart).dispose() line: 487	
XXXMultipageEditor(FormEditor).dispose() line: 405	
XXXMultipageEditor.dispose() line: 202	
CompatibilityEditor(CompatibilityPart).invalidate() line: 260	
CompatibilityEditor(CompatibilityPart).destroy() line: 418	
GeneratedMethodAccessor183.invoke(Object, Object[]) line: not available	
DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: 43	
Method.invoke(Object, Object...) line: 498	
MethodRequestor.execute() line: 58	
InjectorImpl.processAnnotated(Class<Annotation>, Object, Class<?>, PrimaryObjectSupplier, PrimaryObjectSupplier, ArrayList<Class<?>>) line: 1001	
InjectorImpl.processAnnotated(Class<Annotation>, Object, Class<?>, PrimaryObjectSupplier, PrimaryObjectSupplier, ArrayList<Class<?>>) line: 966	
InjectorImpl.uninject(Object, PrimaryObjectSupplier) line: 200	
FieldRequestor(Requestor<L>).uninject(Object, PrimaryObjectSupplier) line: 176	
ContextObjectSupplier$ContextInjectionListener.update(IEclipseContext, int, Object[]) line: 88	
TrackableComputationExt.update(ContextChangeEvent) line: 106	
EclipseContext.removeListenersTo(Object) line: 491	
ContextInjectionFactory.uninject(Object, IEclipseContext) line: 184	
PartRenderingEngine.safeRemoveGui(MUIElement) line: 967	
PartRenderingEngine.access$1(PartRenderingEngine, MUIElement) line: 886	
PartRenderingEngine$4.run() line: 881	
SafeRunner.run(ISafeRunnable) line: 45	
PartRenderingEngine.removeGui(MUIElement) line: 865	
PartRenderingEngine.subscribeTopicToBeRendered(Event) line: 186	
GeneratedMethodAccessor54.invoke(Object, Object[]) line: not available	
DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: 43	
Method.invoke(Object, Object...) line: 498	
MethodRequestor.execute() line: 58	
EventObjectSupplier$DIEventHandler.handleEvent(Event) line: 92	
EventHandlerWrapper.handleEvent(Event, Permission) line: 205	
EventHandlerTracker.dispatchEvent(EventHandlerWrapper, Permission, int, Event) line: 203	
EventHandlerTracker.dispatchEvent(Object, Object, int, Object) line: 1	
EventManager.dispatchEvent(Set<Entry<K,V>>, EventDispatcher<K,V,E>, int, E) line: 234	
ListenerQueue<K,V,E>.dispatchEventSynchronous(int, E) line: 151	
EventAdminImpl.dispatchEvent(Event, boolean) line: 132	
EventAdminImpl.sendEvent(Event) line: 75	
EventComponent.sendEvent(Event) line: 44	
EventBroker.send(String, Object) line: 55	
UIEventPublisher.notifyChanged(Notification) line: 63	
PartImpl(BasicNotifierImpl).eNotify(Notification) line: 424	
PartImpl(UIElementImpl).setToBeRendered(boolean) line: 314	
PartServiceImpl.hidePart(MPart, boolean) line: 1380	
WorkbenchPage.hidePart(MPart, boolean, boolean, boolean, boolean) line: 1558	
WorkbenchPage.hidePart(MPart, boolean, boolean, boolean) line: 1510	
WorkbenchPage.closeEditors(IEditorReference[], boolean) line: 1480	
CloseResourceAction.closeEditors(List<IEditorReference>, IWorkbenchWindow) line: 427	
CloseResourceAction.access$2(List, IWorkbenchWindow) line: 422	
CloseResourceAction$1.run() line: 352	

```

run at the same time

**Comment by cdietrich** [comment]
=> the access to getDescription/the complete descriptionsChanged method and the complete disconnect method are not synchronized.

**Comment by cdietrich** [comment]
Fixed in 2.19
